packages/react-hooks/src/useRefEffect.ts

```diff
@@ -35,31 +35,34 @@ export function useRefEffect<T>(callback: (value: T) => (() => void) | void, ini
     callback: (value: T) => (() => void) | void;
     cleanup?: (() => void) | void;
   };
-
-  const data = React.useRef<RefData>({
-    ref: Object.assign(
-      (value: T | null) => {
-        if (data.ref.current !== value) {
-          if (data.cleanup) {
-            data.cleanup();
-            data.cleanup = undefined;
+  const cleanup = React.useRef<RefData['cleanup']>();
+  const [ref] = React.useState(() => ({
+    // value
+    value: initial,
+    // last callback
+    callback,
+    // "memoized" public interface
+    facade: {
+      get current() {
+        return ref.value;
+      },
+      set current(value) {
+        if (ref.value !== value) {
+          if (cleanup.current) {
+            cleanup.current();
+            cleanup.current = undefined;
           }
 
-          data.ref.current = value;
+          ref.value = value;
 
           if (value !== null) {
-            data.cleanup = data.callback(value);
+            cleanup.current = ref.callback(value);
           }
         }
       },
-      {
-        current: initial,
-      },
-    ),
-    callback,
-  }).current;
-
-  data.callback = callback;
+    },
+  }));
 
-  return data.ref;
+  ref.callback = callback;
+  return (ref.facade as unknown) as RefCallback<T>;
```

Comment:
This is fishy. `facade` is _not_ a `RefCallback` (a function that takes a new value); it's an object with a get/set function for `current`. useRefEffect should be updated to return `MutableRefObject` instead. You'd need to update the places it's used as well...